kubernetes/kubernetes
# Simplify and de-layer Service REST implementation

Simplify Service's REST implementation by using the Begin* hooks introduced in #96393.  This is a large change overall, so the PR is a series of commits that are each designed to be easy to read and review, even if they are incomplete on their own.

Background:
 
Service has an "outer" and "inner" REST handler.  This is because of how we do IP and port allocations synchronously, but since we don't have API transactions, we need to roll those back in case of a failure.  Both layers use the same `Strategy`, but the outer calls into the inner, which causes a lot of complexity in the code (including an open-coded partial reimplementation of a date-unknown snapshot of the generic REST code) and results in `Prepare` and `Validate` hooks being called twice.

This PR aims to merge those layers and use mostly generic REST code.  This DOES NOT attempt to change or fix the way IPs and ports are allocated in the Service REST stack. 

One net result of this is that the validation code *can* get a bit stricter, since it will NEVER see a service before allocation.  This is left for later.

For posterity:

The "normal" REST flow seems to be:

```
mutating webhooks
generic REST store Create {
    cleanup = BeginCreate
    BeforeCreate {
        strategy.PrepareForCreate {
            dropDisabledFields
        }
        strategy.Validate
        strategy.Canonicalize
    }
    createValidation (validating webhooks)
    storage Create
    cleanup
    AfterCreate
    Decorator
}
```

Service (before this PR) does:

```
mutating webhooks
svc custom Create {
    BeforeCreate {
        strategy.PrepareForCreate {
            dropDisabledFields
        }
        strategy.Validate
        strategy.Canonicalize
    }
    Allocations
    inner (generic) Create {
        cleanup = BeginCreate
        BeforeCreate {
            strategy.PrepareForCreate {
                dropDisabledFields
            }
            strategy.Validate
            strategy.Canonicalize
        }
        createValidation (validating webhooks)
        storage Create
        cleanup
        AfterCreate
        Decorator
    }
}
```

After this PR:

```
mutating webhooks
generic REST store Create {
    cleanup = BeginCreate
        Allocations
    BeforeCreate {
        strategy.PrepareForCreate {
            dropDisabledFields
        }
        strategy.Validate
        strategy.Canonicalize
    }
    createValidation (validating webhooks)
    storage Create
    cleanup
    AfterCreate
        Rollback allocations on error
    Decorator
}
```

/kind bug
/kind cleanup

Fixes #87603
Fixes #98994

```release-note
The `Service.spec.ipFamilyPolicy` field is now *required* in order to create or update a Service as dual-stack.  This is a breaking change from the beta behavior.  Previously the server would try to infer the value of that field from either `ipFamilies` or `clusterIPs`, but that caused ambiguity on updates.  Users who want a dual-stack Service MUST specify `ipFamilyPolicy` as either "PreferDualStack" or "RequireDualStack".  
```

cc @khenidak @aojea
FYI @liggitt @deads2k @lavalamp


For followup PR(s):
- Validation overhaul (TODO comments "validation should not pass with empty clusterIP")
  - lots of things verified twice in update path
  - lots of things not verified at all
  - hard to follow
  - tests should state which fields expect errors
  - tests should have a "base" object + list of mutations
- Audit strategy code and tests
  - use svctest
- Test perf (move test boilerplate to outer loops?)
- rename IPFamilyPolicyType and InternalTraffic. - remove Type suffix
- rename ServiceExternalTrafficPolicyTypeLocal -> ServiceTrafficPolicyLocal ?
- rename api.IPv4Protocol
- Logging cleanup and internal Error for IP failures
- make DryRun work the same in NodePort as in IP allocator?
- make dryRun return valid, in-range values ?
- TODO in ResourceLocation()
- Other TODOs
- Scrub allocator code - refactor, renames, comments.
- testcase for dryrun for IPs, NPs, and HCNP (HCNP missing now).  Lean more on {type=LB, ETP=Local} as a case.
- tests for create-on-update
- more feature tests
